BNO.cpp

```diff
@@ -6,57 +6,90 @@
 */
 #include "BNO.h"
 
-LiquidCrystal_I2C lcd(0x27,20,4);
-
 BNO::BNO(){
     bno = Adafruit_BNO055();
 }
 
 double BNO::getDifferenceWithZero() { 
   double error_generated = 0;
   
-  double current_angle_x = getAngleBNOX();
+  double current_angle_x = getAngleX();
   if (current_angle_x >= 180) {
     error_generated = (360 - current_angle_x);
   }
   
   else {
-    error_generated = current_angle;
-    error_generated = -(error_generated);
+    error_generated = -current_angle_x;
   }
   
   return error_generated;
 }
 
-double BNO::getAngleBNOX() { 
+double BNO::getAngleX() { 
   sensors_event_t event;
   bno.getEvent(&event);
   
   return event.orientation.x;
 }
 
-double BNO::getAngleBNOY() { 
+double BNO::getAngleY() { 
   sensors_event_t event;
   bno.getEvent(&event);
   
   return event.orientation.y;
 }
 
-double BNO::getAngleBNOZ() { 
+double BNO::getAngleZ() { 
   sensors_event_t event;
   bno.getEvent(&event);
   
   return event.orientation.z;
 }
 
-void BNO::LCDCalibration() {
-  lcd.init(); 
-  lcd.backlight();
-  lcd.print("Hola Mundo");
-  lcd.setCursor(0, 1);
-  lcd.print(millis()/1000);
-  lcd.print(" Segundos");
-  delay(kTimeToPrint);
+double BNO::getPwmBNOLeft(const double desire) { // Positive.
+    double pwm_left = 0;
+    double pwm_right = 0;
+    double pwm_left_final = 0;
+    double pwm_right_final = 0;
+    double pwm_right_BNO = 0;
+    double pwm_left_BNO = 0;
+    double current_angle_x = getAngleX(); 
+    const double errorBNO = control.getAngleError(current_angle_x, desire);
+
+    if (errorBNO > 0) {
+        pwm_right = motors.kLimit_inf_pwm;
+        pwm_left_BNO = motors.kPAdvance * errorBNO;
+        pwm_left_final = pwm_left_BNO + sensor.getPwmUltrasonicLeft();
+        }
+    else {
+        pwm_left = motors.kLimit_inf_pwm;
+        pwm_right_BNO = motors.kPAdvance * errorBNO;
+        pwm_right_final = pwm_right_BNO + sensor.getPwmUltrasonicRight();;
+    }
+    return pwm_left_final;
+}
+
+double BNO::getPwmBNORight(const double desire) { // Negative.
+    double pwm_left = 0;
+    double pwm_right = 0;
+    double pwm_left_final = 0;
+    double pwm_right_final = 0;
+    double pwm_right_BNO = 0;
+    double pwm_left_BNO = 0;
+    double current_angle_x = getAngleX(); 
+    const double errorBNO = control.getAngleError(current_angle_x, desire);
+
+    if (errorBNO > 0) {
+        pwm_right = motors.kLimit_inf_pwm;
+        pwm_left_BNO = motors.kPAdvance * errorBNO;
+        pwm_left_final = pwm_left_BNO + sensor.getPwmUltrasonicLeft();
+        }
+    else {
+        pwm_left = motors.kLimit_inf_pwm;
+        pwm_right_BNO = motors.kPAdvance * errorBNO;
+        pwm_right_final = pwm_right_BNO + sensor.getPwmUltrasonicRight();;  
+    }
+    return pwm_right_final;
 }
```

Comment:
These two functions should not be in here. They should be in the control part.

In the BNO.h and .cpp only have functions that use ONLY the BNO. The idea is this class to be a wrapper of the actual bno interface, so you can use it easily in other places.